loki/transformations/split_read_write.py

```diff
@@ -0,0 +1,143 @@
+# (C) Copyright 2018- ECMWF.
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation
+# nor does it submit to any jurisdiction.
+
+
+from loki.batch import Transformation, ProcedureItem
+from loki.expression import FindVariables, Array, SubstituteExpressions
+from loki.tools import as_tuple, flatten
+from loki.ir import (
+    pragma_regions_attached, is_loki_pragma, nodes as ir, FindNodes,
+    Transformer, FindScopes
+)
+
+__all__ = ['SplitReadWriteTransformation']
+
+class SplitReadWriteTransformation(Transformation):
+    """
+    When accumulating values to multiple components of an array, a compiler cannot rule out
+    the possibility that the indices alias the same address. Consider for example the following
+    code:
+
+    .. code-block:: fortran
+
+        !$loki split read-write
+        do jlon=1,nproma
+           var(jlon, n1) = var(jlon, n1) + 1.
+           var(jlon, n2) = var(jlon, n2) + 1.
+        enddo
+        !$loki end split read-write
+
+    In the above example, there is no guarantee that ``n1`` and ``n2`` do not in fact point to the same location.
+    Therefore the load and store instructions for ``var`` have to be executed in order.
+
+    For cases where the user knows ``n1`` and ``n2`` indeed represent distinct locations, this transformation
+    provides a pragma assisted mechanism to split the reads and writes, and therefore make the loads independent
+    from the stores. The above code would therefore be transformed to:
+
+    .. code-block:: fortran
+
+        !$loki split read-write
+        do jlon=1,nproma
+           loki_temp_0(jlon) = var(jlon, n1) + 1.
+           loki_temp_1(jlon) = var(jlon, n2) + 1.
+        enddo
+
+        do jlon=1,nproma
+           var(jlon, n1) = loki_temp_0(jlon)
+           var(jlon, n2) = loki_temp_1(jlon)
+        enddo
+        !$loki end split read-write
+    """
+
+    item_filter = (ProcedureItem,)
+
+    def __init__(self, dimensions):
+        self.dimensions = dimensions
+
+    def transform_subroutine(self, routine, **kwargs):
+
+        # initialise working vars, lists and maps
+        temp_vars = []
+        region_map = {}
+        temp_counter = 0
+
+        # cache variable_map for fast lookup later
+        variable_map = routine.variable_map
+
+        # find split read-write pragmas
+        with pragma_regions_attached(routine):
+            for region in FindNodes(ir.PragmaRegion).visit(routine.body):
+                if is_loki_pragma(region.pragma, starts_with='split read-write'):
+
+                    # find assignments inside pragma region
```

Comment:
I feel like this requires quite a lot of IR traversals for something that should be doable in a single pass. For small regions that shouldn't make a big difference, but if there are many then it might pile up. I'm counting at least 6 passes over the region body (in a quick eyeball-estimate).

How about a single tree walk that in-place updates the assignments that need splitting, and collects all assignments at the end?

Drafting here a bit of pseudo code to illustrate what I mean:

```py
class SplitReadWriteWalk(Transformer):

    def __init__(self, **kwargs):
        self.write_assignments = []
        kwargs['inplace'] = True
        super().__init__(**kwargs)

    def visit_Loop(self, o, **kwargs):
        # This collects the relevant loop dimensions in kwargs['dim_nest'] during the
        # tree walk
        dim = [d for d in self.dimensions if d.index == o.variable]
        dim_nest = kwargs.pop('dim_nest', [])
        return super().visit_Node(o, dim_nest=dim_nest + dim, **kwargs)

    def visit_Assignment(self, o, **kwargs):
        if isinstance(o.lhs, Array) and a.lhs.name in o.rhs:
            parent_loop_dims = kwargs.pop('dim_nest', [])
            # <do the logic for shape etc here>
            tmp_var = <...>
            self.write_assignments += [ir.Assignment(lhs=o.lhs, rhs=tmp_var)
            o._update(lhs=tmp_var)
        return o

for region in ...:
    transformer = SplitReadWriteWalk()
    region.body = transformer.visit(region.body)
    if transformer.write_assignments:
        region.body.append(transformer.write_assignments)
```

Not sure if I'm overlooking something, but it feels to me like that should be sufficient and much more efficient.

Reply:
Yes, tree-walk is definitely the way to do it! The use of `FindScopes` to discover a parent loop nest felt a bit clunky to me too but the idea of implementing this as a visitor never occured to me. Many thanks for the suggestion 🙏